File: packages/cli/src/commands/governance.ts
import { Command } from "commander";
import * as path from "node:path";
import fs from "fs-extra";
import { readYamlOrJson } from "../core/schema.js";
import {
  loadAuthorityPolicy,
  explainPath,
  getProtectedPaths,
  checkGovernance,
} from "../core/authority.js";
import {
  defaultChangedFilesSource,
  getGitDiffFiles,
  resolveChangedFiles,
} from "../core/changed-files.js";
import { validate as validateIntervention } from "../validators/intervention.js";

interface GovernanceCheckOptions {
  card?: string;
  diff?: string;
  changedFilesSource?: string;
  enforce?: boolean;
  json?: boolean;
  root?: string;
}

interface GovernanceExplainOptions {
  path?: string;
  json?: boolean;
  root?: string;
}

interface GovernanceListProtectedOptions {
  json?: boolean;
  root?: string;
}

interface InterventionValidateOptions {
  intervention?: string;
  json?: boolean;
}

async function loadCardGovernanceData(
  cardPath: string
): Promise<{ files: string[]; governance?: Record<string, unknown> }> {
  try {
    const data = await readYamlOrJson(cardPath);
    const card = data as Record<string, unknown>;
    const governance = card.governance as Record<string, unknown> | undefined;

    // Extract files_changed from evidence
    const evidence = card.evidence as Record<string, unknown> | undefined;
    if (evidence) {
      const filesChanged = evidence.files_changed as string[] | undefined;
      if (filesChanged && Array.isArray(filesChanged)) {
        return { files: filesChanged, governance };
      }
    }

    return { files: [], governance };
  } catch {
    return { files: [] };
  }
}

export async function governanceCheckAction(
  opts: GovernanceCheckOptions
): Promise<void> {
  const root = path.resolve(opts.root ?? process.cwd());
  let files: string[] = [];
  let cardFiles: string[] = [];
  let gitFiles: string[] = [];
  let changedFilesSource = "card";
  const changedFileErrors: string[] = [];
  const changedFileNotes: string[] = [];
  let governance: Record<string, unknown> | undefined;

  if (opts.card) {
    const cardPath = path.resolve(root, opts.card);
    if (!(await fs.pathExists(cardPath))) {
      console.error(`Error: Card not found: ${cardPath}`);
      process.exit(2);
    }
    const cardData = await loadCardGovernanceData(cardPath);
    cardFiles = cardData.files;
    governance = cardData.governance;
  }

  if (opts.diff && !opts.card) {
    gitFiles = await getGitDiffFiles(opts.diff, root);
    files = gitFiles;
    changedFilesSource = "git";
  } else if (opts.card) {
    const resolved = await resolveChangedFiles({
      cardFiles,
      diffRef: opts.diff,
      root,
      source: defaultChangedFilesSource({
        explicit: opts.changedFilesSource,
        diffRef: opts.diff,
      }),
    });
    files = resolved.files;
    gitFiles = resolved.git_files;
    changedFilesSource = resolved.source;
    changedFileErrors.push(...resolved.errors);
    changedFileNotes.push(...resolved.notes);
  }

  if (changedFileErrors.length > 0) {
    if (opts.json) {
      console.log(
        JSON.stringify(
          {
            ok: false,
            violations: [],
            warnings: [],
            total_violations: 0,
            total_warnings: 0,
            changed_files: {
              source: changedFilesSource,
              card_files: cardFiles,
              git_files: gitFiles,
              files,
            },
            errors: changedFileErrors,
          },
          null,
          2
        )
      );
    } else {
      for (const error of changedFileErrors) console.error(error);
    }
    process.exit(1);
  }

  if (files.length === 0) {
    if (opts.json) {
      console.log(
        JSON.stringify(
          {
            ok: true,
            violations: [],
            warnings: [],
            total_violations: 0,
            total_warnings: 0,
            changed_files: {
              source: changedFilesSource,
              card_files: cardFiles,
              git_files: gitFiles,
              files,
            },
            message: "No files to check",
          },
          null,
          2
        )
      );
    } else {
      console.log("No files to check for governance violations.");
    }
    process.exit(0);
  }

  const result = await checkGovernance(files, root, {
    enforce: opts.enforce === true,
    governance,
  });

  if (opts.json) {
    console.log(
      JSON.stringify(
        {
          ok: result.total_violations === 0 && result.total_warnings === 0,
          violations: result.violations,
          warnings: result.warnings,
          total_violations: result.total_violations,
          total_warnings: result.total_warnings,
          report_only: result.report_only,
          enforced: result.enforced,
          changed_files: {
            source: changedFilesSource,
            card_files: cardFiles,
            git_files: gitFiles,
            files,
          },
          notes: changedFileNotes,
        },
        null,
        2
      )
    );
  } else {
    if (result.violations.length > 0) {
      console.log("Authority violations (enforced mode):");
      for (const violation of result.violations) {
        console.log(`  [${violation.authority}] ${violation.path}`);
        console.log(`    ${violation.rationale}`);
        if (violation.approval_note) {
          console.log(`    ${violation.approval_note}`);
        }
      }
      console.log("");
      console.log(`Total: ${result.total_violations} violation(s)`);
    } else if (result.warnings.length > 0) {
      console.log("Authority warnings (report-only, no admission block):");
      for (const warning of result.warnings) {
        console.log(`  [${warning.authority}] ${warning.path}`);
        console.log(`    ${warning.rationale}`);
      }
      console.log("");
      console.log(
        `Total: ${result.total_warnings} warning(s) - admission NOT blocked (PR2 report-only)`
      );
    } else {
      console.log("No governance violations found.");
    }
  }

  process.exit(result.total_violations > 0 ? 1 : 0);
}

export async function governanceExplainAction(
  opts: GovernanceExplainOptions
): Promise<void> {
  if (!opts.path) {
    console.error("Error: --path is required");
    process.exit(2);
  }

  const root = path.resolve(opts.root ?? process.cwd());
  const targetPath = path.isAbsolute(opts.path)
    ? opts.path
    : path.resolve(root, opts.path);

  try {
    const result = await explainPath(targetPath, root);

    if (opts.json) {
      console.log(
        JSON.stringify(
          {
            path: result.path,
            authority: result.authority,
            rationale: result.rationale,
          },
          null,
          2
        )
      );
    } else {
      console.log(`Path: ${result.path}`);
      console.log(`Authority: ${result.authority}`);
      console.log(`Rationale: ${result.rationale}`);
    }
  } catch (err) {
    console.error(`Error: ${err instanceof Error ? err.message : String(err)}`);
    process.exit(2);
  }
}

export async function governanceListProtectedAction(
  opts: GovernanceListProtectedOptions
): Promise<void> {
  const root = path.resolve(opts.root ?? process.cwd());

  try {
    const policy = await loadAuthorityPolicy(root);
    const protectedPaths = getProtectedPaths(policy);

    if (opts.json) {
      console.log(
        JSON.stringify(
          {
            authority_classes: policy.authority_classes,
            protected_paths: protectedPaths,
          },
          null,
          2
        )
      );
    } else {
      console.log("Authority classes:");
      for (const [name, cls] of Object.entries(policy.authority_classes)) {
        console.log(`  ${name}: ${cls.description}`);
      }
      console.log("");
      console.log("Protected paths:");
      for (const pp of protectedPaths) {
        console.log(`  ${pp.path} -> ${pp.authority}`);
        console.log(`    ${pp.rationale}`);
      }
    }
  } catch (err) {
    console.error(`Error: ${err instanceof Error ? err.message : String(err)}`);
    process.exit(2);
  }
}

export async function interventionValidateAction(
  opts: InterventionValidateOptions
): Promise<void> {
  if (!opts.intervention) {
    console.error("Error: --intervention <path> is required");
    process.exit(2);
  }

  const interventionPath = path.resolve(opts.intervention);

  if (!(await fs.pathExists(interventionPath))) {
    console.error(`Error: Intervention file not found: ${interventionPath}`);
    process.exit(2);
  }

  try {
    const data = await readYamlOrJson(interventionPath);
    const result = await validateIntervention(data);

    if (opts.json) {
      console.log(
        JSON.stringify(
          {
            valid: result.valid,
            errors: result.errors,
          },
          null,
          2
        )
      );
    } else {
      if (result.valid) {
        console.log("Intervention is valid.");
      } else {
        console.log("Intervention validation failed:");
        for (const error of result.errors) {
          console.log(`  - ${error}`);
        }
      }
    }

    process.exit(result.valid ? 0 : 1);
  } catch (err) {
    console.error(`Error: ${err instanceof Error ? err.message : String(err)}`);
    process.exit(2);
  }
}

export function governanceCommand(): Command {
  const cmd = new Command("governance").description(
    "Governance boundary and authority checking (report-only, no admission block)"
  );

  cmd
    .command("check")
    .description(
      "Check files against authority boundary (report-only warnings, no admission block)"
    )
    .option("--card <path>", "Path to completion card to check files from")
    .option(
      "--diff <ref>",
      "Git ref to diff against (e.g., HEAD) to get changed files"
    )
    .option(
      "--changed-files-source <mode>",
      "Changed files source when --card and --diff are both set: card, git, union, strict"
    )
    .option(
      "--enforce",
      "Exit non-zero for unapproved authority violations",
      false
    )
    .option("--json", "Output JSON instead of human-readable text", false)
    .option("--root <path>", "Repository root", process.cwd())
    .action(governanceCheckAction);

  cmd
    .command("explain")
    .description("Explain authority class for a given path")
    .option("--path <path>", "Path to explain authority for")
    .option("--json", "Output JSON instead of human-readable text", false)
    .option("--root <path>", "Repository root", process.cwd())
    .action(governanceExplainAction);

  cmd
    .command("list-protected")
    .description("List all protected paths and their authority classes")
    .option("--json", "Output JSON instead of human-readable text", false)
    .option("--root <path>", "Repository root", process.cwd())
    .action(governanceListProtectedAction);

  return cmd;
}

export function interventionCommand(): Command {
  const cmd = new Command("intervention").description(
    "Intervention artifact validation"
  );

  cmd
    .command("validate")
    .description("Validate an intervention artifact against the schema")
    .option("--intervention <path>", "Path to intervention YAML/JSON file")
    .option("--json", "Output JSON instead of human-readable text", false)
    .action(interventionValidateAction);

  return cmd;
}

File: internal/cli/governance.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/BrianNguyen29/x-harness/internal/authority"
)

func handleGovernance(args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, "governance requires a subcommand: check, explain, list-protected")
		return ExitUsage
	}

	switch args[0] {
	case "check":
		return handleGovernanceCheck(args[1:], stdout, stderr)
	case "explain":
		return handleGovernanceExplain(args[1:], stdout, stderr)
	case "list-protected":
		return handleGovernanceListProtected(args[1:], stdout, stderr)
	case "-h", "--help", "help":
		fmt.Fprintln(stderr, "usage: x-harness governance <check|explain|list-protected> [options]")
		return ExitUsage
	default:
		fmt.Fprintf(stderr, "unknown governance subcommand: %s\n", args[0])
		fmt.Fprintln(stderr, "usage: x-harness governance <check|explain|list-protected> [options]")
		return ExitUsage
	}
}

func parseGovernanceFlags(args []string, requireCard bool, stderr io.Writer) (card, diff, changedFilesSource, root string, enforce, jsonMode bool, exitCode int) {
	root = "."
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--card":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --card requires a value")
				return "", "", "", "", false, false, ExitUsage
			}
			card = args[i+1]
			i++
		case "--diff":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --diff requires a value")
				return "", "", "", "", false, false, ExitUsage
			}
			diff = args[i+1]
			i++
		case "--changed-files-source":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --changed-files-source requires a value")
				return "", "", "", "", false, false, ExitUsage
			}
			changedFilesSource = args[i+1]
			i++
		case "--enforce":
			enforce = true
		case "--json":
			jsonMode = true
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return "", "", "", "", false, false, ExitUsage
			}
			root = args[i+1]
			i++
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return "", "", "", "", false, false, ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return "", "", "", "", false, false, ExitUsage
		}
	}

	if requireCard && card == "" {
		fmt.Fprintln(stderr, "Error: --card is required")
		return "", "", "", "", false, false, ExitUsage
	}

	root, _ = filepath.Abs(root)
	return card, diff, changedFilesSource, root, enforce, jsonMode, -1
}

func handleGovernanceCheck(args []string, stdout, stderr io.Writer) int {
	card, _, changedFilesSource, root, enforce, jsonMode, exitCode := parseGovernanceFlags(args, true, stderr)
	if exitCode >= 0 {
		return exitCode
	}

	cardPath := card
	if !filepath.IsAbs(cardPath) {
		cardPath = filepath.Join(root, card)
	}
	if _, err := os.Stat(cardPath); os.IsNotExist(err) {
		fmt.Fprintf(stderr, "Error: Card not found: %s\n", cardPath)
		return ExitError
	}

	files, governance, err := authority.LoadCardGovernanceData(cardPath)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	// Simple changed-files-source support: if source is git, ignore card files
	// For simplest viable parity, we primarily use card files.
	// If source is explicitly "git" and no files, that's an error scenario we handle gracefully.
	if changedFilesSource == "git" {
		files = []string{}
	}

	if len(files) == 0 {
		if jsonMode {
			out := map[string]any{
				"ok":                true,
				"violations":        []any{},
				"warnings":          []any{},
				"total_violations":  0,
				"total_warnings":    0,
				"changed_files":     files,
				"message":           "No files to check",
			}
			data, _ := json.MarshalIndent(out, "", "  ")
			fmt.Fprintln(stdout, string(data))
		} else {
			fmt.Fprintln(stdout, "No files to check for governance violations.")
		}
		return ExitOK
	}

	policy, err := authority.LoadAuthorityPolicy(root)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	result, err := authority.CheckGovernance(files, root, policy, &authority.GovernanceCheckOptions{
		Enforce:    enforce,
		Governance: governance,
	})
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		out := map[string]any{
			"ok":                result.TotalViolations == 0 && result.TotalWarnings == 0,
			"violations":        result.Violations,
			"warnings":          result.Warnings,
			"total_violations":  result.TotalViolations,
			"total_warnings":    result.TotalWarnings,
			"report_only":       result.ReportOnly,
			"enforced":          result.Enforced,
			"changed_files":     files,
			"notes":             []string{},
		}
		data, _ := json.MarshalIndent(out, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		if result.TotalViolations > 0 {
			fmt.Fprintln(stdout, "Authority violations (enforced mode):")
			for _, v := range result.Violations {
				fmt.Fprintf(stdout, "  [%s] %s\n", v.Authority, v.Path)
				fmt.Fprintf(stdout, "    %s\n", v.Rationale)
				if v.ApprovalNote != "" {
					fmt.Fprintf(stdout, "    %s\n", v.ApprovalNote)
				}
			}
			fmt.Fprintln(stdout, "")
			fmt.Fprintf(stdout, "Total: %d violation(s)\n", result.TotalViolations)
		} else if result.TotalWarnings > 0 {
			fmt.Fprintln(stdout, "Authority warnings (report-only, no admission block):")
			for _, w := range result.Warnings {
				fmt.Fprintf(stdout, "  [%s] %s\n", w.Authority, w.Path)
				fmt.Fprintf(stdout, "    %s\n", w.Rationale)
			}
			fmt.Fprintln(stdout, "")
			fmt.Fprintf(stdout, "Total: %d warning(s) - admission NOT blocked (PR2 report-only)\n", result.TotalWarnings)
		} else {
			fmt.Fprintln(stdout, "No governance violations found.")
		}
	}

	if result.TotalViolations > 0 {
		return ExitError
	}
	return ExitOK
}

func handleGovernanceExplain(args []string, stdout, stderr io.Writer) int {
	pathFlag := ""
	root := "."
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--path":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --path requires a value")
				return ExitUsage
			}
			pathFlag = args[i+1]
			i++
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return ExitUsage
			}
			root = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if pathFlag == "" {
		fmt.Fprintln(stderr, "Error: --path is required")
		return ExitUsage
	}

	root, _ = filepath.Abs(root)

	policy, err := authority.LoadAuthorityPolicy(root)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	result, err := authority.ExplainPath(policy, pathFlag, root)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		data, _ := json.MarshalIndent(result, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintf(stdout, "Path: %s\n", result.Path)
		fmt.Fprintf(stdout, "Authority: %s\n", result.Authority)
		fmt.Fprintf(stdout, "Rationale: %s\n", result.Rationale)
	}

	return ExitOK
}

func handleGovernanceListProtected(args []string, stdout, stderr io.Writer) int {
	root := "."
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return ExitUsage
			}
			root = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	root, _ = filepath.Abs(root)

	policy, err := authority.LoadAuthorityPolicy(root)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return ExitError
	}

	protectedPaths := authority.GetProtectedPaths(policy)

	if jsonMode {
		out := map[string]any{
			"authority_classes": policy.AuthorityClasses,
			"protected_paths":   protectedPaths,
		}
		data, _ := json.MarshalIndent(out, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintln(stdout, "Authority classes:")
		for name, cls := range policy.AuthorityClasses {
			fmt.Fprintf(stdout, "  %s: %s\n", name, cls.Description)
		}
		fmt.Fprintln(stdout, "")
		fmt.Fprintln(stdout, "Protected paths:")
		for _, pp := range protectedPaths {
			fmt.Fprintf(stdout, "  %s -> %s\n", pp.Path, pp.Authority)
			fmt.Fprintf(stdout, "    %s\n", pp.Rationale)
		}
	}

	return ExitOK
}
